runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler.cpp

```diff
@@ -0,0 +1,151 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openopenj9.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "j9protos.h"
+#include "j9vmnls.h"
+#include "BytecodeAction.hpp"
+#include "LayoutFFITypeHelpers.hpp"
+#include "OutOfLineINL.hpp"
+#include "VMHelpers.hpp"
+#include "UnsafeAPI.hpp"
+
+extern "C" {
+
+#if JAVA_SPEC_VERSION >= 16
+/**
+ * openj9.internal.foreign.abi.InternalUpcallHandler: private native long allocateUpcallStub(UpcallMHMetaData mhMetaData, String[] cSignatureStrs);
+ *
+ * @brief Request the JIT to generate an upcall thunk of the specified java method
+ *
+ * @param mhMetaData[in] a data object that consists of the upcall handler and the metadata for MH resolution
+ * @param cSignatureStrs[in] An array of the native signature string for the requested java method during the upcall
+ * @return the address of the native thunk generated by JIT.
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler_allocateUpcallStub(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9UpcallMetaData *upcallMetaData = NULL;
+	j9object_t mhMetaData = NULL;
+	J9UpcallNativeSignature *nativeSig = NULL;
+	J9UpcallSigType *sigArray = NULL;
+	void *thunkAddr = NULL;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	j9object_t cSigStrs = J9_JNI_UNWRAP_REFERENCE(currentThread->sp);
```

Comment:
I think you really want `(j9object_t)currentThread->sp[0]` here and similar below. The macro essentially does this, but its use is confusing.

Reply:
Agreed and replaced with `(j9object_t)currentThread->sp[0]`.